Console over websocket

EmilyxFox opened the pull request:
## What
I'm trying to add a websocket console similar to the ssh console that's already there.

This was also requested in https://github.com/itzg/docker-minecraft-server/issues/3152.

## Things that currently work:
- connect to ws
- stream stdout & err over the websocket
- send stdin over websocket

## Things that don't work
- ~~sending the last x amount of lines to newly connected clients~~
  - Works!
- requesting older lines (for scrolling up in a console, for example)
- ~~websocket isn't gracefully closed by the server before the program exits~~ 
  - ~~closes with 1006~~
  - closes with 1001! 🎉

## Potential features
https://github.com/itzg/docker-minecraft-server/issues/3152 mentions being able to request things like available commands, auto completions, and various server stats (cpu, ram). I don't know the feasibility of this at all but I thought I'd include it here anyways just to track.

I think being able to query server status (starting, running, stopping) would be cool, or having it sent in a heartbeat type thing.

## Why
Feature could be useful for making web panel consoles which I think would be cool as a sidecar.

I think it would also be a good idea to have two formats (json & text) to either use json formatted messages with extra stuff in, or just send plain stdout/err and stdin through the websocket.

I haven't tested a lot of edge cases at all since this code is an absolute mess. I did test the really long line bug though, and it seems to handle that fine. I make no promises that I'll finish this, but I've put it here to gather some feedback :)

EmilyxFox committed:
server to client working

EmilyxFox committed:
client to server working

strongdoctor commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,256 @@
+			Content: string([]byte(msg)),
+			Time:    time.Now(),
+		}
+		// err := client.c.Write(ctx, websocket.MessageText, )
```
DELET

EmilyxFox replied:
👍

EmilyxFox committed:
deleted cringe

EmilyxFox commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,256 @@
+	MessageTypeStdout  MessageType = "stdout"
+	MessageTypeStderr  MessageType = "stderr"
+	MessageTypeWelcome MessageType = "welcome"
+	// MessageTypeAuthErr MessageType = "auth_err"
```
keeping this cus should probably add auth before this is done but i added a comment to the comment hope you like it

itzg reviewed (commented):
Sorry for the delay reviewing this. Overall, I really like this, especially the new "multi" abstractions.

EmilyxFox committed:
fix unexported variable names

EmilyxFox committed:
fix passing in context

EmilyxFox committed:
pass in context to echo

EmilyxFox committed:
pass in stderrWriter

EmilyxFox commented on main.go:
```diff
@@ -73,18 +74,34 @@ func main() {
+		StdoutWritersList = append(StdoutWritersList, wsOutWriter)
+		StderrWritersList = append(StderrWritersList, wsErrWriter)
+
+		go runWebsocketServer(logger, wsOutWriter, stdin)
```
:+1:

EmilyxFox committed:
remove unused import

EmilyxFox committed:
better variable names

EmilyxFox committed:
put log back in its place

EmilyxFox commented on main.go:
```diff
@@ -97,14 +114,19 @@ func main() {
+	cmd.Stderr = multiErr
 
+	if !args.RemoteConsole {
+		logger.Debug("Directly assigning stdin")
```
probably just lost the overview :)

EmilyxFox committed:
remove comments

EmilyxFox commented:
Actually sending stderr works now (tested). I think I should remove the prefix option since its useless with the json output.

EmilyxFox commented:
@itzg I'd be interested to hear your take on the potential features bit. Especially which ones are kinda doable and which ones are less so.

itzg commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,255 @@
+	logger  *zap.Logger
+	stdin   io.Writer
+	clients map[uuid.UUID]*WsClient
+	mu      sync.Mutex
```
Yeah, don't worry about it. As you can see, I talked myself out of it by the end 😄

itzg commented on remote_shell_service.go:
```diff
@@ -456,3 +452,43 @@ func runRemoteShellServer(console *Console, logger *zap.Logger) {
+	// Write to all pipe writers
+	for _, w := range pw.writers {
+		if _, err := w.Write(p); err != nil {
+			// Handle error - maybe remove this writer or log
```
Oh yeah, the zap logger has to be propagated as an instance field. That's overall a good thing, but annoying in edge cases like this.

Zap itself isn't the challenge -- it's more about passing through the logger instance from the code that creates the pipe writer.

EmilyxFox replied:
I was kinda thinking that I couldn't pass the logger into the `Write` method since it wouldn't fit the shape of the `io.Writer` interface anymore or something along those lines hopefully my confused ramblings make sense

itzg replied:
Sorry, I didn't mean passing it via the Write call but rather as a field of `pipeWriter` to be used within the `Write` method. That field could be set via a parameter of `newPipeWriter`.

itzg commented:
> @itzg I'd be interested to hear your take on the potential features bit. Especially which ones are kinda doable and which ones are less so.

It seems like the ones you noted, available commands, auto completions, and various server stats (cpu, ram), are not particularly doable without deeper hooks into with a plugin/mod at which point that plugin/mod should be the one exposing a websocket interface. 

My expectations for this websocket support are covered since it can do the equivalent of rcon's sending commands and returning the response. The prior and current logging retrieval was even a bonus/optional capability in my mind.

Reporting server up/down status does seem doable and helpful via a call to mc-monitor.

EmilyxFox commented:
I think authentication might still be a good idea before its merged. It could just be a header or query param or something like that, I think, and then just taken from the rcon password.

itzg commented:
Ah, great point about authentication, but yes keeping it simple like leveraging the rcon password as a bearer token, etc seems good.

EmilyxFox committed:
add authentication

EmilyxFox commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,290 @@
+}
+
+func (s *websocketServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	password := r.Header.Get("X-WS-Auth")
+	if password != getWebsocketPassword() {
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusUnauthorized)
+
+		errMsg := AuthErrorMessage{
+			Type:   MessageTypeAuthError,
+			Reason: "invalid password",
+		}
+		json.NewEncoder(w).Encode(errMsg)
+		return
+	}
```
I would like to be able to skip this with `WEBSOCKET_DISABLE_AUTHENTICATION=true` but I would have to coerce "true" "TRUE" and maybe "1" to bools and idk you might have a fancy way of doing that otherwise this works. lmk if you don't like the header name or whatnot

itzg replied:
I'd recommend adding it to `Args` like you did the other new field:

https://github.com/itzg/mc-server-runner/pull/124/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R22

but adding in a `WithEnv("")`, [documented here](https://pkg.go.dev/github.com/itzg/go-flagsfiller#hdr-Environment_variable_mapping), in the flags parse call at

https://github.com/itzg/mc-server-runner/pull/124/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R45

EmilyxFox commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,290 @@
+		return password
+	}
+
+	password = os.Getenv("RCON_PASSWORD")
```
I'm sort of assuming that the autogenerated rcon password is passed along with this environment variable since that's (how the ssh bit gets it as well)[https://github.com/itzg/mc-server-runner/blob/f76cec2bf69aa9dbc5f572523a4ca9e624847de8/remote_shell_service.go#L105].

itzg replied:
Yep, at this point `RCON_PASSWORD` is loaded in as an env var (via the shell's `export` function) here:

https://github.com/itzg/docker-minecraft-server/blob/99f31d5a9b2e6a31e0dc170c637650881ade2a6a/scripts/start-configuration#L78-L79

That will then get inherited by the mc-server-runner execution.

itzg commented:
> (I can only reference changes in comments) I'm wondering if finally closing the websocket just before [this done](https://github.com/itzg/mc-server-runner/pull/124/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261L195) (or maybe after) would be the best place to do it? You'd want to get the whole shutdown sequence sent over the websocket.

Yes but even better is to hook into the context that gets cancelled there. Only quirk is that would mean moving the context creation here

https://github.com/itzg/mc-server-runner/pull/124/files#diff-2873f79a86c0d8b3335cd7731b0ecf7dd4301eb19a82ef7a1cba7589b5252261R153

further up and passing that into the websocket server object setup.

I'm likely over simplifying the suggestion so I don't mind pushing a change to show what I mean.

EmilyxFox commented:
Feel free to push your suggestion. Its quite late here, so no rush, I probably wont write any more code today :saluting_face:

EmilyxFox committed:
enable the disabling of authentication 🤓

EmilyxFox committed:
remove message prefixes

EmilyxFox committed:
add error logging to ssh pipe writer

EmilyxFox committed:
websocket disable authentication false by default

itzg committed:
Tie main context and err chan into websocket server

itzg commented:
Hopefully this wasn't too disruptive and so far only confirmed the startup/shutdown behavior

https://github.com/itzg/mc-server-runner/pull/124/commits/8be524ad606814c65aead74b78f1a2ee9bf5d680

EmilyxFox commented:
I just went to test this and it doesn't seem to be closing the websockets. I tried with a shell script and when that exits it doesn't close the websocket and same with a minecraft server that was sent "stop".

itzg commented:
Sorry, yeah I only tested the term signal handling and forgot to check for "server" stopping. 

The exit handling in the runner had already gotten more convoluted than I usually prefer, so maybe I should just revert most of what I was trying there. 

I'll investigate just a little first with the server exiting scenario.

itzg committed:
Added WaitGroup to coordinate websocket shutdown

EmilyxFox commented:
Mojang is adding a feature very similar to this to their server, and I think some things should be reconsidered in that context.
The new management server is likewise a websocket server that lets you receive events and perform actions on the server. The features of the Mojang management server and the remote console don't overlap, and actually complement each other quite nicely.
Due to the similarity of these two features, I think we should consider whether to try and piggy-back off their features, or try to distance them as much as possible. Being somewhere in the middle is the worst option in my opinion.
In 25w37a they added various authorisation and security features. First, an authorisation bearer token. This is similar to the simple authorisation that I added to this feature, but perhaps just slightly different enough for it to be confusing. (Different header name, secret isn't prefixed with "bearer", secret is derived from RCON password and not server management secret). They also added TLS and various settings related to that. That's been a goal of this feature as well, and I think it should be similarly considered whether to use certificates provided to the management server by default.

That was kind of a vague argument _for_ using management server settings with the websocket console. I think an argument against could be that it could be confusing that there's an official Mojang management server and a remote console provided by the image that both use the same settings. It might also be an idea to wait a little while with merging this feature to see what Mojang settles on (mTLS for example). I'd be super interested in feedback on these thoughts!

itzg commented:
Building on those thoughts, I'm wondering if we should wait and see what else they add in general. It would be both cool and sad if they end up adding everything you have implemented in your websocket interface. 

I'm now wondering if they have a place where they solicit ideas for that specifically. It seems like they have gotten more closed off to direct idea submissions where it started out in Reddit. 

If it settles as is, I think I'm leaning towards the idea of reusing the same auth material as theirs.

EmilyxFox commented:
> I'm wondering if we should wait and see what else they add in general.

I think that's a good idea. I highly doubt that they'll make the console available over json-rpc, but who knows.
> It seems like they have gotten more closed off to direct idea submissions where it started out in Reddit.

I seemingly can't even read bug reports without making an account!
> If it settles as is, I think I'm leaning towards the idea of reusing the same auth material as theirs.

Alright. I might start working in that direction then. I kinda got a bit stuck on closing the ws connections before stopping the program. Your commits of cancelling the context doesn't seem to work still, but I wanted to test it with another client.

EmilyxFox committed:
change echo function name to handleIncoming

EmilyxFox committed:
add log lines ring buffer

EmilyxFox commented:
I added a pretty simple ring buffer for recent log lines, but I don't think its particularly thread safe. I think skipping log lines, or getting duplicates is totally possible. I think I'll make recent lines optional for the client

EmilyxFox committed:
idea for making recent lines optional
it could also be a path or a query param

EmilyxFox committed:
fix typo in WelcomeMessage struct..
these should probably not be capitalised

EmilyxFox committed:
clerify WebsocketAddress arg

EmilyxFox committed:
stop logging error if client goes away

EmilyxFox committed:
remove header to request log history

EmilyxFox committed:
use container/ring for log history

EmilyxFox committed:
change auth to use Sec-WebSocket-Protocol

EmilyxFox commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,360 @@
+	disableAuth bool
+}
+
+type logLinesRingBuffer struct {
```
👍

EmilyxFox committed:
remove closed connections from WsClients map

EmilyxFox committed:
more extensible error handling

EmilyxFox committed:
add origin check

EmilyxFox committed:
remove capitalisation from unexported things

EmilyxFox committed:
send correct message type (stdin/stderr) in websocket

EmilyxFox commented:
What do you think of pulling all this out into a separate package? The file is getting a bit long. No problem if not, but it might be a bit tidier

EmilyxFox committed:
add arg for log buffer size

itzg commented:
> What do you think of pulling all this out into a separate package? The file is getting a bit long. No problem if not, but it might be a bit tidier

Would be good if it's easy enough, but I don't feel strongly either way.

itzg commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,441 @@
+				}
+				s.stdin.Write([]byte(content))
+			}
+		}
```
Would it be worth logging a warning for unexpected message types?

itzg reviewed (commented):
Sorry it toot a while to follow up to latest changes. Looking great. Just a couple very minor things. Otherwise, move it out of draft when you're ready for final review and merge.

EmilyxFox committed:
remove commented time/rate import

EmilyxFox committed:
prepare for more message types

EmilyxFox committed:
break up function call

EmilyxFox commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,441 @@
+	"github.com/coder/websocket/wsjson"
+	"github.com/google/uuid"
+	"go.uber.org/zap"
+	// "golang.org/x/time/rate"
```
👍

EmilyxFox commented on websocket_shell_service.go:
```diff
@@ -0,0 +1,441 @@
+	}
+}
+
+func runWebsocketServer(ctx context.Context, logger *zap.Logger, errorChan chan error, finished *sync.WaitGroup, stdoutWriter *wsWriter, stderrWriter *wsWriter, stdin io.Writer, disableAuth bool, address string, trustedOrigins []string, disableOriginCheck bool, logBufferSize int) {
```
👍

EmilyxFox committed:
change content field to data

EmilyxFox committed:
change trusted origins to allowed origins to match mojang management server

Mqxx commented:
I tried something similar since I opened this issue (https://github.com/itzg/docker-minecraft-server/issues/3152). I have written [a mod for subscribing to different channels over websocket](https://github.com/MinecraftPlayground/websocket-console-mod) the idea behind this was that you would be able to "connect" so specified "endpoints" to receive logging information, like so:

```
localhost:8080/log?channel=debug -> DEBUG Channel
localhost:8080/log?channel=info -> INFO Channel
localhost:8080/log?channel=warn -> WARN Channel
localhost:8080/log?channel=error -> ERROR Channel

localhost:8080/log -> localhost:8080/log?channel=debug&channel=info&channel=warn&channel=error -> All Channels

localhost:8080/log?channel=warn&channel=error -> WARN, ERROR Channel
```

Sending is done using a HTTP POST:
```
POST localhost:8080/command

body {
  command: "setblock 0 0 0 minecraft:diamond_block"
}
```

I don't know how your current implementation for this is, but maybe it is useful for you 😊, at least that's how I would implement it.

itzg commented:
> But my question would be if it is "best practice" to also use the WebSocket connection for stdin

Yes it is. See the docker attach websocket API I posted above. 

If you don't do them in the same stream then you're back to a challenge of coordinating of sending a command one way and getting output of that command another.

EmilyxFox commented:
> > Yes it is. See the docker attach websocket API I posted above.
> 
> Okay, but the initial connection (as stated by the docs) is also just a simple HTTP POST with the *option* to upgrade. This does not directly has something to do with "best practices" for "bi-directional" communication.
> 
> Like I said, you can do it that way, don't get me wrong, but I would split these into two separate APIs. One for just streaming the logs (because there I really need live updates), and one for sending commands, get configurations, get stats or something else. I think of this wohle concept not only like Minecraft server management, but rather Minecraft server **Container** management. And this includes much more than just "sending commands" and/or receiving "logs".

The functionality you describe here is how RCON works afaik. By just having access to stdin and stdout, I don't think it's possible to link a command to its output. I might just not be thinking of the right way, but to me it seems impossible. I do see the utility in having a POST endpoint for sending commands, but you couldn't reliably get the output back. RCON is probably the tool for the job here. @itzg also already has a library for that 🫡

itzg commented:
> By just having access to stdin and stdout, I don't think it's possible to link a command to its output.

Good point when comparing to Rcon where they precisely tag the request ID with the response content. 

At least with both stdin and stdout over the same websocket, as you've done, then the relative sequence of input and output messages is significant and useful.

Mqxx commented:
> But how would you link an input to a specific output?

You don't. And you can't. Thats my point. But you can give a specific "feedback" for each command send to the server stdin. You can not do that over WebSocket, at least not without an proper protocol like JsonRPC 2.0

Mqxx commented:
But I am not familiar with go-lang that's why I am not telling you how to do it and instead offering solutions how you *could* build it and make it (in my opinion) better 😊

Mqxx commented:
> I think doing it any other way would be unnecessarily worse than RCON, but then the solution itself seems like a needless abstraction on top of RCON.

No abstraction can be worst than RCON itself 😅, also I don't think it would be a "needless abstraction"

Mqxx commented:
I would suggest, to not get to much out of scope to implement the sending and receiving using only the WebSocket for now with a subpath (`/console`) and expand that in the future.

EmilyxFox committed:
make use of args for websocket password

EmilyxFox committed:
uniform logging style

EmilyxFox committed:
change struct fields to be more descriptive

EmilyxFox committed:
change "authErr" to "authFailure"
apparently failure is more descriptive that its the client's fault and not an error in the auth process

EmilyxFox committed:
put endpoint on `/console`

EmilyxFox committed:
remove unnecessary time fields

EmilyxFox committed:
go mod tidy

EmilyxFox committed:
update readme

EmilyxFox committed:
don't log stack trace on json parse error

EmilyxFox committed:
close with 1001 when server stops

Mqxx commented:
@EmilyxFox @itzg I cooked something in the meantime: https://github.com/MinecraftPlayground/server-management-mod/wiki

Maybe you can get some inspiration from that.

itzg commented:
I'm sorry I basically forgot about this PR. 

@Mqxx what endpoint are those operations sending to? The ones being added in this PR?

itzg reviewed (approved):
Sorry for losing this in the pile. Continues to look great.

itzg commented:
This is now released in https://github.com/itzg/mc-server-runner/releases/tag/1.14.0
